tls/s2n_server_cert_verify.c

```diff
@@ -13,14 +13,141 @@
  * permissions and limitations under the License.
  */
 
+#include "tls/s2n_certificate_verify.h"
+#include "tls/s2n_connection.h"
+#include "crypto/s2n_hash.h"
+
+#include "stuffer/s2n_stuffer.h"
 #include "error/s2n_errno.h"
+#include "utils/s2n_safety.h"
+
+static int s2n_server_write_cert_verify_signature(struct s2n_connection *conn, struct s2n_stuffer *out);
+static uint8_t s2n_server_cert_verify_header_length();
 
 int s2n_server_cert_verify_send(struct s2n_connection *conn)
 {
-    S2N_ERROR(S2N_ERR_UNIMPLEMENTED);
+    struct s2n_stuffer *out = &conn->handshake.io;
+
+    /* Write the algorithm to out */
+    GUARD(s2n_stuffer_write_uint8(out, conn->secure.conn_hash_alg));
+    GUARD(s2n_stuffer_write_uint8(out, conn->secure.conn_sig_alg));
+
+    /* Write digital signature */
+    GUARD(s2n_server_write_cert_verify_signature(conn, out));
+
+    return 0;
 }
 
 int s2n_server_cert_verify_recv(struct s2n_connection *conn)
 {
-    S2N_ERROR(S2N_ERR_UNIMPLEMENTED);
+    struct s2n_stuffer *in = &conn->handshake.io;
+    DEFER_CLEANUP(struct s2n_blob signed_content = {0}, s2n_free);
+    uint8_t hash_algorithm;
+    uint8_t sign_algorithm;
+    uint8_t unsigned_content_length;
+    uint8_t hash_digest_length;
+    uint16_t signature_size;
+
+    DEFER_CLEANUP(struct s2n_hash_state message_hash = {0}, s2n_hash_free);
+    DEFER_CLEANUP(struct s2n_hash_state hash_copy = {0}, s2n_hash_free);
+    uint8_t digest_out[S2N_MAX_DIGEST_LEN];
+    GUARD(s2n_hash_new(&message_hash));
+    GUARD(s2n_hash_new(&hash_copy));
+
+    /* Read the algorithm */
+    GUARD(s2n_stuffer_read_uint8(in, &hash_algorithm));
+    GUARD(s2n_stuffer_read_uint8(in, &sign_algorithm));
+    S2N_ERROR_IF(hash_algorithm != conn->secure.conn_hash_alg, S2N_ERR_BAD_MESSAGE);
```

Comment:
I'm not sure if `.conn_hash_alg` and `.conn_sig_alg` at this time in setup in s2n handshake, we should add a task to track it if we haven't

Reply:
maybe my assumption was incorrect, but this message occurs immediately before the finish, so i thought they would have been

Reply:
confirmed in stand-up. these should be set by this point

Reply:
had a chat with @raycoll about this. `conn_hash_alg` and `conn_sig_alg ` would be set after client_hello in server mode but not in client mode where s2n_server_cert_verify_recv() is called. This is somewhat related to #1257. There may be a few solutions but the easy one would be to relax these 2 lines of checks and the more proper solution would be to use `s2n_get_signature_hash_pair_if_supported(sig_hash_algs, &hash_algorithm, &signature_algorithm))`